src/sc/sidechainTxsCommitmentGuard.h

```diff
@@ -0,0 +1,94 @@
+#ifndef SIDECHAIN_TX_COMMITMENT_GUARD
+#define SIDECHAIN_TX_COMMITMENT_GUARD
+
+#include "coins.h"
+#include <sc/sidechaintypes.h>
+
+class uint256;
+class CTransaction;
+
+class CScCertificate;
+class CTxForwardTransferOut;
+class CBwtRequestOut;
+class CTxCeasedSidechainWithdrawalInput;
+
+class SidechainTxsCommitmentGuard
+{
+public:
+    SidechainTxsCommitmentGuard() {};
+    ~SidechainTxsCommitmentGuard() {};
+
+    SidechainTxsCommitmentGuard(const SidechainTxsCommitmentGuard&) = delete;
+    SidechainTxsCommitmentGuard& operator=(const SidechainTxsCommitmentGuard&) = delete;
+
+    bool add(const CTransaction& tx, bool autoRewind = false);
+    bool add(const CScCertificate& cert);
+
+    void rewind(const CTransaction& tx);
+    void rewind(const CScCertificate& cert);
+
+private:
+    bool add_fwt(const CTxForwardTransferOut& ccout);
+    bool add_bwtr(const CBwtRequestOut& ccout);
+    bool add_csw(const CTxCeasedSidechainWithdrawalInput& ccin);
+    bool add_cert(const CScCertificate& cert);
+
+    void rewind(const CTransaction& tx, const int failingFt, const int failingBwtr, const int failingCsw);
+    void keepMapsClean();
+
+    // Keeping information separated to closely mimic CCTPlib structures
+    struct CommitmentBuilderStatsAliveCounter {
+        uint32_t   ft = 0;
+        uint32_t bwtr = 0;
+        uint32_t cert = 0;
+        uint32_t  bwt = 0;
+    };
+    struct CommitmentBuilderStatsCeasedCounter {
+        uint32_t  csw = 0;
+    };
+    struct CommitmentBuilderStats {
+        std::map<uint256, CommitmentBuilderStatsAliveCounter>  cbsaMap;
+        std::map<uint256, CommitmentBuilderStatsCeasedCounter> cbscMap;
+        // The following values MUST be aligned with those specified in CCTPlib!
+        static const int   SC_LIMIT = 4096;
```

Comment:
Just a note: I've carefully double-checked the value for this limit, 4096 seems safe.
But in general, can we get all these constants from `zendoo-mc-cryptolib`?

Reply:
Thanks for checking! Getting the limits directly from mc-crypto or cctplib was the original idea, but unfortunately they require a rework on both libraries: their symbols are defined in cctplib, but not exposed to mc-crypto.
We plan to do this in another PR, however, to simplify the alignment of these limits.